[SwiftUI] Introduce WebView

rr-codes opened the pull request:
#### 8f4a11e48e28bdf25e9e111c7d211d15404cdffb
<pre>
[SwiftUI] Introduce WebView
<a href="https://bugs.webkit.org/show_bug.cgi?id=283464">https://bugs.webkit.org/show_bug.cgi?id=283464</a>
<a href="https://rdar.apple.com/140322360">rdar://140322360</a>

Reviewed by Wenson Hsieh.

Add a SwiftUI view which connects to WebPage to present the web content.

* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.backingWebView):
* Source/WebKit/UIProcess/API/Swift/WebView.swift: Copied from Tools/MiniBrowser/AppDelegate.swift.
(WebView_v0.body):
(WebViewRepresentable.makePlatformView(_:)):
(WebViewRepresentable.updatePlatformView(_:context:)):
(WebViewRepresentable.makeUIView(_:)):
(WebViewRepresentable.updateUIView(_:context:)):
(WebViewRepresentable.makeNSView(_:)):
(WebViewRepresentable.updateNSView(_:context:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/MiniBrowser/AppDelegate.swift:
(BrowserAppDelegate.createSwiftUIWindow(_:)):
* Tools/MiniBrowser/Configurations/Base.xcconfig:
* Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj:
* Tools/MiniBrowser/SwiftUI/BrowserViewModel.swift: Copied from Tools/MiniBrowser/AppDelegate.swift.
(BrowserViewModel.displayedURL):
(BrowserViewModel.didReceiveNavigationEvent(_:)):
(BrowserViewModel.navigateToSubmittedURL):
* Tools/MiniBrowser/SwiftUI/ContentView.swift: Added.
(ContentView.urlFieldIsFocused):
(ContentView.body):
* Tools/MiniBrowser/SwiftUI/MiniBrowserView.swift: Copied from Tools/MiniBrowser/AppDelegate.swift.
(MiniBrowserView.body):

Canonical link: <a href="https://commits.webkit.org/286964@main">https://commits.webkit.org/286964@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc829cca7eadfaf93e0c20f000d17bde407361c3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77623 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56657 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30546 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82260 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28902 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79740 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65813 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4957 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60838 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18788 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80689 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50797 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66621 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41110 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48189 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24116 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27229 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69302 "Found 1 new API test failure: TestWebKitAPI.WebKit.FindInUnifiedPDFAfterFindInPage (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24459 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83616 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5004 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3403 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69046 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5161 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66583 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68310 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12315 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10424 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12036 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4951 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7763 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4966 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8404 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6729 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

axiixc commented on Tools/MiniBrowser/SwiftUI/BrowserViewModel.swift:
```diff
@@ -0,0 +1,56 @@
+    var displayedURL: String = ""
+
+    func didReceiveNavigationEvent(_ event: WebPage.NavigationEvent) {
+        Self.logger.info("Did receive navigation event \(String(describing: event.kind)) for navigation \(String(describing: event.navigationID))")
```
(For later) Perhaps we should consider supplying a `CustomStringConvertible` conformance to avoid the `String(describing:)`

aprotyas commented on Source/WebKit/UIProcess/API/Swift/WebView.swift:
```diff
@@ -0,0 +1,75 @@
+        owner.page.backingWebView
+    }
+
+    func updatePlatformView(_ nsView: WKWebView, context: Context) {
```
Nit: maybe call this argument `platformView`?

aprotyas replied:
(and let the argument for `updateUIView` be `uiView`?)

rr-codes replied:
ooh yes, this was a copy and paste mistake. thanks!

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebView.swift:
```diff
@@ -0,0 +1,75 @@
+    let owner: WebView_v0
+
+    func makePlatformView(context: Context) -> WKWebView {
+        owner.page.backingWebView
```
This might not be robust enough long term. We'll need to figure out what happens when a `WebPage` moves between representable. We can't have multiple owners regardless, but we'll want to decide if it's an error, if we can handle it gracefully, and how deterministic it might even be. Like perhaps we should keep an ownership assertion which we can tear down in something like `dismantleUIView()`.

I think in any case having a wrapper view used as the representable's own view is probably more flexible, and then we can parent the web view in something like `update*` if it changes.

axiixc commented on Tools/MiniBrowser/SwiftUI/ContentView.swift:
```diff
@@ -0,0 +1,81 @@
+            .toolbar {
+                ToolbarItemGroup(placement: .principal) {
+                    VStack(spacing: 0) {
+                        @Bindable var viewModel = viewModel
```
I can't remember, does `@Environment` give you a bindable value already?

rr-codes replied:
it does not (somewhere in the docs it says to do exactly this, I forget exactly where though)

axiixc commented on Tools/MiniBrowser/SwiftUI/ContentView.swift:
```diff
@@ -0,0 +1,81 @@
+                            .focused($urlFieldIsFocused)
+
+                        ProgressView(value: viewModel.page.estimatedProgress, total: 1.0)
+                            .tint(.accentColor)
```
Isn't this the default?

axiixc commented on Tools/MiniBrowser/SwiftUI/BrowserViewModel.swift:
```diff
@@ -0,0 +1,56 @@
+
+    private static let logger = Logger(subsystem: Bundle.main.bundleIdentifier!, category: String(describing: BrowserViewModel.self))
+
+    var page = WebPage()
```
```suggestion
    let page = WebPage()
```

axiixc commented on Tools/MiniBrowser/SwiftUI/ContentView.swift:
```diff
@@ -0,0 +1,81 @@
+                            .textFieldStyle(.roundedBorder)
+                            .focused($urlFieldIsFocused)
+
+                        ProgressView(value: viewModel.page.estimatedProgress, total: 1.0)
```
This is kinda the "it's hard to remember to always use `self.`" issue that has led most people to drop it unless required.

axiixc reviewed (commented):
Looks good to me

rr-codes committed:
[SwiftUI] Introduce WebView
https://bugs.webkit.org/show_bug.cgi?id=283464
rdar://140322360

Reviewed by Wenson Hsieh.

Add a SwiftUI view which connects to WebPage to present the web content.

* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.backingWebView):
* Source/WebKit/UIProcess/API/Swift/WebView.swift: Copied from Tools/MiniBrowser/AppDelegate.swift.
(WebView_v0.body):
(WebViewRepresentable.makePlatformView(_:)):
(WebViewRepresentable.updatePlatformView(_:context:)):
(WebViewRepresentable.makeUIView(_:)):
(WebViewRepresentable.updateUIView(_:context:)):
(WebViewRepresentable.makeNSView(_:)):
(WebViewRepresentable.updateNSView(_:context:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/MiniBrowser/AppDelegate.swift:
(BrowserAppDelegate.createSwiftUIWindow(_:)):
* Tools/MiniBrowser/Configurations/Base.xcconfig:
* Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj:
* Tools/MiniBrowser/SwiftUI/BrowserViewModel.swift: Copied from Tools/MiniBrowser/AppDelegate.swift.
(BrowserViewModel.displayedURL):
(BrowserViewModel.didReceiveNavigationEvent(_:)):
(BrowserViewModel.navigateToSubmittedURL):
* Tools/MiniBrowser/SwiftUI/ContentView.swift: Added.
(ContentView.urlFieldIsFocused):
(ContentView.body):
* Tools/MiniBrowser/SwiftUI/MiniBrowserView.swift: Copied from Tools/MiniBrowser/AppDelegate.swift.
(MiniBrowserView.body):

Canonical link: https://commits.webkit.org/286964@main
